New logging layer

ktsaou opened the pull request:
- [x] support structured logs
- [x] support systemd-journal for netdata logs
- [x] file format changed to logfmt
- [x] file format dates changed to iso8601 UTC
- [x] structured annotations for HTTP/HTTPS requests
- [x] structured annotations for ACLK requests
- [x] structured annotations for WebRTC requests
- [x] structured annotations for STREAM RECEIVERS
- [x] structured annotations for PLUGINS.D
- [x] structured annotations for internal plugins
- [x] structured annotations for health alert transitions 

---

- [x] added command `systemd-cat-native` that receives a stream formatted in the [systemd-native protocol](https://systemd.io/JOURNAL_NATIVE_PROTOCOL/) and sends all log messages to journal. This command is installed in `/usr/sbin`, much like Netdata itself.
- [x] added command `log2journal` that receives log lines and converts them to systemd journal export format.

---

@Ferroin this PR adds tools that will be enabled only when `libcurl` and `libpcre2-8` are available during compilation, and in case of dynamic linking, during runtime too.

I have modified `netdata.spec` to include `systemd-cat-native` which should be available anywhere Netdata is available. It does not require any additional dependencies and libcurl is optional. So, if Netdata is available on any system, this tool should be too. It will greatly benefit from `libcurl`, so please make sure we ship packages that depend on `libcurl`.

I have not added `log2journal` in `netdata.spec`. This tool is standalone. It does not link with libnetdata. It only requires `libpcre2-8`. It won't be compiled if this library is missing. So, I don't know how to add it to `netdata.spec`.

Both tools should be available everywhere. Especially for `systemd-cat-native`, this tool is required by Netdata external plugins, alert notifications and scripts. So, it MUST be always available, even if the system is not systemd based, or even not Linux. When systemd is not available, it falls back to log-files or syslog, the same way Netdata does. So, despite its name, it is an integral part of Netdata, that MUST me available everywhere.

thiagoftsm commented:
Compiling on `Slackware current`, I have this warning:

```sh
exporting/aws_kinesis/aws_kinesis.c: In function 'aws_kinesis_connector_worker':
./libnetdata/log/log.h:225:143: warning: unknown escape sequence: '\040'
  225 | #define netdata_log_debug(type, args...) do { if(unlikely(debug_flags & type)) netdata_logger(NDLS_DEBUG, NDLP_DEBUG, __FILE__, __FUNCTION__, __LINE__, ##args); } while(0)
      |                                                                                                                                               ^~~~~~~~
exporting/aws_kinesis/aws_kinesis.c:155:13: note: in expansion of macro 'netdata_log_debug'
  155 |             netdata_log_debug(D_EXPORTING,
      |             ^~~~~~~~~~~~~~~~~
```

underhood commented on aclk/aclk.c:
```diff
@@ -408,7 +408,7 @@ void aclk_graceful_disconnect(mqtt_wss_client client)
         }
     }
     netdata_log_info("ACLK link is down");
```
any reason this remains `netdata_log_info`

ktsaou replied:
I have not replaced everything. Old log lines are still valid, they are mapped to the new logging system.

ktsaou commented:
I have also added documentation in `README.md` for you to see what you can configure.

ktsaou commented:
Debug.log will be removed forever. No need for it.

Collector.log and error.log now go to journal when there is a journal in the system.

ktsaou commented:
And even without journal, error.log is renamed to daemon.log

thiagoftsm commented:
Thank you to explain @ktsaou, but I do not have after last commits a log file named `daemon.log` in `/var/log/netdata`, neither on Slackware  Current nor on FreeBSD. I used `mlocate` to search on the whole system, and I cannot see in any other directory.

@Ferroin, considering that you could recreate the previous issue on your FreeBSD installation, can you confirm to us that you only have `health.log` and `aclk.log`?

![logs](https://github.com/netdata/netdata/assets/49162938/8f747afc-e726-4a14-a59e-5af8707741c5)

ktsaou commented:
@thiagoftsm as I said, `daemon.log` and `collector.log` goes to journal on systemd-journal systems. You will not have it. Check `netdata.conf`. It should say `journal`, not a filename.

ilyam8 commented:
> @thiagoftsm as I said, `daemon.log` and `collector.log` goes to journal on systemd-journal systems. You will not have it. Check `netdata.conf`. It should say `journal`, not a filename.

@ktsaou, @thiagoftsm is correct, his VM is FreeBSD (systemdless system). It should have daemon/collector.log files. I installed ND on FreeBSD and see the same issue:

```bash
[cm@freebsd-vm /opt/netdata/var/log/netdata]$ ls -l
total 45
-rw-rw-r--  1 netdata  netdata       0 Nov 15 01:08 aclk.log
-rw-rw-r--  1 netdata  netdata  228923 Nov 15 01:12 health.log
```

ktsaou commented:
Found the issue. Accidentally, all file logs were going to aclk.log. Fixed it.
Also fixed a couple more bugs.

Now health.log is only about alert transitions and goes to journal by default.
So, currently only access.log and aclk.log are outside journal.

ktsaou commented:
@ilyam8 I have updated systemd-journal plugin. Now some fields are exposed to systemd-journal by default.

thiagoftsm commented:
> Found the issue. Accidentally, all file logs were going to aclk.log. Fixed it. Also fixed a couple more bugs.

I can confirm the issues are fixed, and we have logs again on environments that do not have systemd (Slackware current and FreeBSD). Thank you!

ilyam8 commented:
Latest version doesn't set the `NETDATA_LOG_SEVERITY_LEVEL` env variable. This variable is used by all external plugins/helper binaries (e.g. cgroup-network) and bash scripts (e.g. cgroup-name.sh).

ktsaou commented:
I will export both, just in case.

ilyam8 commented:
Not very important information, but related to this PR: journald by default forwards logs to the syslog ([ForwardToSyslog](https://www.freedesktop.org/software/systemd/man/latest/journald.conf.html#ForwardToSyslog =)), so all logs will be stored in `/var/log/syslog` and `/var/log/daemon.log` too.

tkatsoulas commented on collectors/cgroups.plugin/cgroup-network-helper.sh:
```diff
@@ -29,65 +29,117 @@
 
-[ $(( BASH_VERSINFO[0] )) -lt 4 ] && \
-    fatal "BASH version 4 or later is required (this is ${BASH_VERSION})."
+if [ ${BASH_VERSINFO[0]} -lt 4 ]; then
```
```suggestion
if [ "${BASH_VERSINFO[0]}" -lt 4 ]; then
```

tkatsoulas replied:
when we make sure that the tests are ok for https://github.com/netdata/netdata/pull/16357/commits/e57a2574bcb51a9a0b906fc18c6700e936f4de71
don't forger to merge also  this suggestion (forgot it)

tkatsoulas commented:
Add a to-do list: 

TODO:
- [ ] Add dependencies when building from source
